### ISSUE TITLE
fix(metrics): LOC min/max merges child aggregate instead of child min/max — nested granularity lost

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#437]
## Summary

`Loc` min/max aggregation folds the child space's *aggregate value*
(`other.sloc()`) into the parent min/max instead of the child's own
min/max (`other.sloc_min`/`sloc_max`) — unlike every other metric, which folds
`other.X_min`/`other.X_max`. Combined with the guarded `compute_minmax`, this
loses the granularity of deeply nested function spaces, so per-function LOC
min/max is reported over top-level spaces only.

## Location

- `src/metrics/loc.rs:100-103` (`Sloc::merge`), `:159-167` (`Ploc::merge`),
  `:227-234` (`Cloc::merge`), `:288-295` (`Lloc::merge`), `:404-405` (blank in `Stats::merge`)
- Contrast: `src/metrics/cyclomatic.rs:106-107`, `cognitive.rs:100-101`,
  `exit.rs:83-84`, `nargs.rs:105-108`, `nom.rs:110-113`, `tokens.rs:92-93`,
  `abc.rs:200-205` (all fold `other.X_min`/`X_max`)

## Evidence

```rust
// loc.rs — folds the child's aggregate value, NOT its min/max
pub fn merge(&mut self, other: &Sloc) {
    self.sloc_min = self.sloc_min.min(other.sloc() as usize);   // other.sloc(), not other.sloc_min
    self.sloc_max = self.sloc_max.max(other.sloc() as usize);
}
pub(crate) fn compute_minmax(&mut self) {
    if self.sloc_min == usize::MAX {            // guard: once a child folded, skip own value
        self.sloc_min = self.sloc_min.min(self.sloc() as usize);
        self.sloc_max = self.sloc_max.max(self.sloc() as usize);
    }
}
```

```rust
// cyclomatic.rs — every sibling folds the child's min/max
self.cyclomatic_min = self.cyclomatic_min.min(other.cyclomatic_min);
```

Worked example — file (Unit) → class `C` (span 6) → method `m` (sloc 3):
1. `m` leaf `compute_minmax`: `sloc_min = 3`.
2. `C.merge(m)`: `C.sloc_min = min(MAX, m.sloc()=3) = 3`.
3. `C.compute_minmax`: guard sees `sloc_min != MAX` → skips C's own span 6.
4. `unit.merge(C)`: `unit.sloc_min = min(MAX, C.sloc()=6) = 6` — **discards**
   `C.sloc_min = 3`.

Result: file-level `sloc_min = 6`, when the smallest function leaf is `3`. The
sibling metrics fold `other.X_min` and carry `3` correctly. Same defect for
`ploc`/`lloc`/`cloc`/`blank`.

Secondary: `Stats::merge` blank (`:404-405`) folds `other.blank() as usize`;
`blank()` (`sloc - ploc - only_comment_lines`) is an unguarded `f64` subtraction
that can go negative, in which case `as usize` saturates to `0` and corrupts
`blank_min` toward 0 (the serialized `blank()` can also be negative).

## Expected Behavior

`Loc` merge should fold `other.sloc_min`/`other.sloc_max` (and the
ploc/lloc/cloc/blank equivalents), matching all sibling metrics, so nested
function-space min/max propagate to the root. The intended "min over which
spaces?" convention (whether container base values are included — see the
related observation below) should be applied uniformly with the other metrics.

## Actual Behavior

LOC per-function min/max reflects only top-level spaces, dropping nested
function leaves; blank can corrupt toward 0 or serialize negative.

## Impact

Incorrect per-function LOC min/max for any file containing classes with methods
(or nested functions) — i.e. most real-world OO code.

## Related observation (decide + document, not necessarily a separate fix)

`cyclomatic`/`cognitive`/`exit`/`nargs`/`abc`/`tokens`/`nom` `compute_minmax`
fold `self.X` unconditionally (so the Unit base and container bases are included,
e.g. `cyclomatic_min` ≈ always 1), whereas LOC's guarded `compute_minmax`
excludes containers. The two families answer "min over which spaces?"
differently; whichever convention is intended should be applied uniformly and
documented.

---

## Resolution (fix/issue-437)

**Convention chosen:** LOC now matches the sibling metrics — `compute_minmax`
folds each space's own value **unconditionally**, so containers (Unit, class,
namespace, impl) participate in min/max, and `merge` folds `other.X_min` /
`other.X_max` so nested leaves propagate to the root. LOC's prior
guarded-exclude-containers behaviour is removed; the new convention is
documented inline at every `merge` / `compute_minmax` site in
`src/metrics/loc.rs`.

**Secondary (negative blank):** `blank()` is clamped at `0.0`, so the
`sloc - ploc - only_comment_lines` subtraction can no longer serialize a
negative value or corrupt `blank_min` via an `as usize` saturation at the
merge site.

Regression tests added for Rust/Java/Python nested-space propagation and for
non-negative blank; loc and integration snapshots refreshed (value-only).
Status: fixed, not yet released.

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

`Loc` `merge` folds the child's aggregate value (`other.sloc()`) into the
parent's min/max, whereas every sibling metric folds the child's `min`/`max`. The
guarded `compute_minmax` then prevents containers from re-seeding their own
value, so nested function leaves are dropped above the first level.

### Implementation Steps

1. In `src/metrics/loc.rs`, change each `merge` (`Sloc`/`Ploc`/`Cloc`/`Lloc` and
   the blank handling in `Stats::merge`) to fold `other.sloc_min`/`other.sloc_max`
   (and the corresponding `*_min`/`*_max`) instead of `other.sloc()`.
2. Reconcile `compute_minmax` with the chosen convention so leaf seeding still
   happens exactly once (mirror the unconditional-`self.X` fold used by the
   sibling metrics, or keep the guard but seed from the child min/max).
3. Guard the `blank()` subtraction against going negative before `as usize`.
4. Decide and document the "min over which spaces?" convention (container/unit
   base inclusion) so LOC and the other metrics agree.

### Tests to Add/Update

- A nested file→class→method fixture asserting file-level `sloc_min` equals the
  smallest function leaf (e.g. 3), not the container span (6). Same for
  ploc/lloc/cloc.
- A fixture where `blank()` would be negative, asserting it clamps to 0.
- Expect snapshot shifts for LOC min/max on nested files — review the diff is
  min/max-only.

### Verification

- [ ] `cargo test -p big-code-analysis` passes (`--all-features`)
- [ ] `cargo clippy -p big-code-analysis` clean
- [ ] Manual: `bca -m` on an OO file reports per-function LOC min from the
      smallest method.

**Comment by dekobon** [comment]
## Resolution

Fixed in `fix(metrics): propagate nested-space LOC min/max` on branch `fix/issue-437`.

### What was wrong
`Loc` `merge` folded each child space's *aggregate* value (`other.sloc()`,
`other.ploc()`, …) into the parent's min/max instead of the child's own
`min`/`max`, while the guarded `compute_minmax` skipped a container's own
span once any child had folded. Together these dropped nested function-space
granularity, so per-function LOC min/max reflected only top-level spaces
(the worked example's `file -> class C -> method m` reported `sloc_min = 6`
instead of `3`).

### The fix
- `merge` now folds `other.X_min` / `other.X_max` for `sloc`, `ploc`, `cloc`,
  `lloc`, and `blank` — matching every sibling metric (cyclomatic, cognitive,
  exit, nargs, nom, tokens, abc).
- `compute_minmax` folds the space's own value **unconditionally** (the
  `usize::MAX` guard is removed).

### Convention decision (the "min over which spaces?" question)
**LOC now matches the siblings: containers are included.** Each space runs
`compute_minmax` (folding its own value) before being merged upward, and
`merge` folds the child min/max — so the Unit base and every container
(class, namespace, impl) participate in min/max, and nested leaves propagate
to the root. This is the most defensible choice for cross-metric consistency,
and removes LOC's prior deliberate-but-undocumented deviation. The convention
is documented inline at each `merge` / `compute_minmax` site in `src/metrics/loc.rs`.

### Secondary: negative `blank()`
`blank()` = `sloc - ploc - only_comment_lines` is an f64 subtraction that
could go negative; the `as usize` cast at the merge site saturated a negative
to `0` and corrupted `blank_min`. `blank()` is now clamped at `0.0`
(`.max(0.0)`), so the serialized value can no longer be negative and
`blank_min` is not corrupted. The clamp lives in `blank()` itself so every
consumer (serialization, average, merge, compute_minmax) sees the clamped value.

### Tests
- Added nested-leaf min/max propagation tests for Rust, Java, and Python
  (`*_nested_min_max_propagates`): assert `sloc_min` is the smallest leaf
  method span (not the unit span), `sloc_max` is the largest space, and
  `lloc_min` reflects the smallest leaf. Verified via revert that these
  **fail** against the pre-fix merge (reporting the unit span, e.g. `10.0`
  instead of `3.0`).
- Added `rust_blank_never_negative` asserting `blank()`, `blank_min()`,
  `blank_max()` are all `>= 0.0`.
- Refreshed the loc inline/file snapshots and the `big-code-analysis-output`
  integration snapshots; all diffs are value-only on loc min/max/blank fields
  with no structural (key/row/column) changes.
